src/dnx.coreclr.unix/dnx.coreclr.cpp

```diff
@@ -0,0 +1,227 @@
+    }
+
+    if (ppszTpaAssemblyNames != NULL)
+    {
+        FreeTpaBase(ppszTpaAssemblyNames, cTpaAssemblyNames);
+    }
+
+    return true;
+}
+
+// TODO: Figure out if runtimeDirectory should have a trailing "/".  Need to look at what happens on Windows.
+void* LoadCoreClr(std::string& runtimeDirectory)
+{
+    void* ret = nullptr;
+
+    char* coreClrEnvVar = getenv("CORECLR_DIR");
+
+    if (coreClrEnvVar != nullptr)
+    {
+        runtimeDirectory = std::string("CORECLR_DIR");
+
+        std::string coreClrDllPath = runtimeDirectory;
+        coreClrDllPath.append("/");
+        coreClrDllPath.append("libcoreclr.so");
+
+        ret = dlopen(coreClrDllPath.c_str(), RTLD_NOW | RTLD_GLOBAL);
+    }
+
+    if (ret == nullptr)
+    {
+        // Try to load coreclr from application path.
+        char pathToBootstrapper[PATH_MAX];
+        ssize_t pathLen = readlink("/proc/self/exe", pathToBootstrapper, PATH_MAX - 1);
+
+        if (pathLen != -1)
+        {
+            pathToBootstrapper[pathLen] = '\0';
+            runtimeDirectory.assign(pathToBootstrapper);
+
+            size_t lastSlash = runtimeDirectory.rfind('/');
+            runtimeDirectory.erase(lastSlash);
+
+            std::string coreClrDllPath = runtimeDirectory;
+            coreClrDllPath.append("/");
+            coreClrDllPath.append("libcoreclr.so");
+
+            ret = dlopen(coreClrDllPath.c_str(), RTLD_NOW | RTLD_GLOBAL);
+        }
+    }
+
+    return ret;
+}
+
+extern "C" HRESULT CallApplicationMain(PCALL_APPLICATION_MAIN_DATA data)
+{
+    HRESULT hr = S_OK;
+    size_t cchTrustedPlatformAssemblies = 0;
+    std::string runtimeDirectory;
+
+    if (data->runtimeDirectory != NULL)
+    {
+        runtimeDirectory = std::string(data->runtimeDirectory);
+    }
+    else
+    {
+        // TODO: This should get the directory that this library is in, not the CWD.
+        char szCurrentDirectory[PATH_MAX];
+
+        if (getcwd(szCurrentDirectory, PATH_MAX) == NULL)
+        {
+            return E_FAIL;
+        }
+
+        runtimeDirectory = std::string(szCurrentDirectory);
+    }
+
+    std::string coreClrDirectory;
+    void* coreClr = LoadCoreClr(coreClrDirectory);
+
+    if (coreClr == nullptr)
+    {
+        char* error = dlerror();
+        fprintf(stderr, "failed to locate coreclr.dll with error %s\n", error);
+        return E_FAIL;
+    }
+
+    const char* property_keys[] =
+    {
+        "APPBASE",
+        "TRUSTED_PLATFORM_ASSEMBLIES",
+        "APP_PATHS",
+    };
+
+    std::string trustedPlatformAssemblies;
+
+    // Try native images first
+    if (!GetTrustedPlatformAssembliesList(coreClrDirectory.c_str(), true, trustedPlatformAssemblies))
+    {
+        if (!GetTrustedPlatformAssembliesList(coreClrDirectory.c_str(), false, trustedPlatformAssemblies))
+        {
+            fprintf(stderr, "Failed to find files in the coreclr directory\n");
+            hr = E_FAIL;
+            return hr;
+            // goto Finished;
```

Comment:
Ahh, yes.  I will remove this line.  It's vestigial code from a previous attempt at structuring this file.  Thanks!